### DISCUSSION TITLE
Cannot tab-navigate between TextInput widgets on MacOS

### DISCUSSION THREAD

**Discussion by baryonyx5** [beeware/toga#1375] [Q&A]
I just started experimenting with toga this weekend and am seeing an odd behavior on MacOS. I'm not sure if this is a bug or if I'm doing something wrong.

The snippet below reproduces it. Essentially, I have a two column layout where an action in the left column (button press in this example) displays a set of input widgets in the right column.

Launch app:
    The right box shows a set of labels and text inputs
    The first text input widget has focus
    Pressing tab key selects the next text input widget

Press a 'Toggle Forms' button    
    The right box shows a new set of labels/inputs
    None of the new inputs has focus
    After focusing one of the inputs, tab key does not navigate to next widget

I would expect to be able to tab between the widgets once one was in focus.  It does work as expected on Windows 10. 

I'm using toga  0.3.0.dev29 on both platforms. MacOS version is 11.6.1

```python
import toga
from toga.constants import COLUMN, ROW
from toga.style import Pack


style = Pack(flex=1, direction=COLUMN, padding=10)


def divider():
    return toga.Divider(style=Pack(padding_top=12, padding_bottom=12))


def get_fields(id_):
    return [
        toga.Label(f"Form {id_} Input 1", style=Pack(padding=10)),
        toga.TextInput(style=Pack(padding=10)),
        toga.Label(f"Form {id_} Input 2", style=Pack(padding=10)),
        toga.TextInput(style=Pack(padding=10)),
        toga.Label(f"Form {id_} Input 3", style=Pack(padding=10)),
        toga.TextInput(style=Pack(padding=10)),
    ]


class TextInputApp(toga.App):

    def on_replace_box(self, *args, **kwargs):
        """
        Display a new set of input widgets by creating a new form Box
        with new children
        """
        id_ = 2 if self.active_id == 1 else 1
        self.box.remove(self.form_box)
        self.form_box = toga.Box(children=get_fields(id_), style=style)

        self.box.add(self.form_box)
        self.active_id = id_

    def on_replace_children(self, *args, **kwargs):
        """
        Display a new set of input widgets by replacing the children
        on the existing form Box
        """
        id_ = 2 if self.active_id == 1 else 1
        fields = get_fields(id_)
        self.form_box.remove(*self.form_box.children)
        self.form_box.add(*fields)
        self.active_id = id_

    def on_focus(self, *args, **kwargs):
        input2 = self.form_box.children[3]
        input2.focus()

    def startup(self):
        self.active_id = 1
        self.main_window = toga.MainWindow(title=self.name)

        lbl1 = toga.Label("Remove, Recreate Box")
        btn1 = toga.Button("Toggle Forms", style=Pack(flex=1), on_press=self.on_replace_box)
        lbl2 = toga.Label("Replace Children")
        btn2 = toga.Button("Toggle Forms", style=Pack(flex=1), on_press=self.on_replace_children)
        lbl3 = toga.Label("Focus Input 2")
        btn3 = toga.Button("Focus Input 2", style=Pack(flex=1), on_press=self.on_focus)
        self.btn_box = toga.Box(
            children=[lbl1, btn1, divider(), lbl2, btn2, divider(), lbl3, btn3],
            style=style,
        )

        fields = get_fields(self.active_id)
        self.form_box = toga.Box(children=fields, style=style)
        self.box = toga.Box(children=[self.btn_box, self.form_box], style=Pack(direction=ROW))
        self.main_window.content = self.box
        self.main_window.show()


def main():
    return TextInputApp('TextInput', 'org.beeware.widgets.textinput')


if __name__ == '__main__':
    app = main()
    app.main_loop()
```

Worth mentioning that if I reuse the same widget instances instead of recreating them each time (i.e `self.fields1 = get_fields(1)` `self.fields2 = get_fields(2)` and toggle between these in the callbacks), then the same problem occurs when toggling to the second set of widgets but the focus and tab navigation work when toggling back to the widgets initially loaded. 

I hope that makes sense.  

Thanks

**Comment by freakboy3742** [comment]
Thanks for the report - I can't see anything obviously wrong with the code you've provided, so it definitely sounds like a bug. My guess is that it will be related to the dynamic creation of widgets, but I'd need to go digging to elaborate beyond that.
